src/armscan_env/wrapper.py

```diff
@@ -128,26 +164,116 @@ def __getattr__(self, item: str) -> Any:
+                f"Observation spaces are not of type Box: {type(observation)}, {type(additional_obs)}",
+            ) from None
+        return full_obs
+
+
+class AddRewardDetailsWrapper(AddObservationsWrapper):
+    @property
+    def additional_obs_space(self) -> gym.spaces:
+        return self._additional_obs_space
+
+    def __init__(
+        self,
+        env: LabelmapEnv | Env[ObsType, ActType],
+        num_steps_to_observe: int | None = None,
+        additional_obs: Observation = ActionRewardObservation().to_array_observation(),
+    ):
+        """Adds the action that would lead to the highest image variance to the observation.
+        In focus-stigmation agents, this helps in the initial exploratory phase of episodes, as it
+        allows wandering around the state space without worrying about losing track of the
+        best image found so far.
+
+        :param env:
+        :param num_steps_to_observe: Number of steps to observe to pick the highest reward state.
+            If None, all steps are observed.
+        """
+        self._additional_obs_space = additional_obs.observation_space
+        # don't move above, see comment in AddObservationsWrapper
+        super().__init__(env, additional_obs)
+        self.num_steps_to_observe = num_steps_to_observe
+
+        self.reset_wrapper()
+
+    def reset_wrapper(self) -> None:
+        if self.num_steps_to_observe is None:
+            self.rewards: list[float] = []
+            self.states: list[LabelmapStateAction] = []
+        else:
+            self.rewards = deque(maxlen=self.num_steps_to_observe)  # type: ignore
+            self.states = deque(maxlen=self.num_steps_to_observe)  # type: ignore
+
+    def reset(self, **kwargs: Any) -> tuple[ObsType, dict[str, Any]]:
+        self.reset_wrapper()
+        obs, info = super().reset(**kwargs)
+        updated_obs = cast(ObsType, self.observation(obs))
+        return updated_obs, info
+
+    def step(
+        self,
+        action: ActType,
+    ) -> tuple[ObsType, SupportsFloat, bool, bool, dict[str, Any]]:
+        obs, reward, terminated, truncated, info = self.env.step(action)
+        updated_obs = cast(ObsType, self.observation(obs))
+        return updated_obs, reward, terminated, truncated, info
+
+    def _update_observation_fields(self) -> None:
+        tissue_clusters = TissueClusters.from_labelmap_slice(
+            self.env.cur_state_action.labels_2d_slice,
+        )
+        clustering_reward = anatomy_based_rwd(tissue_clusters=tissue_clusters)
+        self.rewards.append(clustering_reward)
+        self.states.append(self.env.cur_state_action)
+        self.highest_rew_state_arr = self.states[np.argmax(self.rewards)]
+
+    def get_additional_obs(self) -> np.ndarray:
+        # base_obs is not used, instead we directly access the current image from the env
+        self._update_observation_fields()
+        return self.additional_obs.compute_observation(self.highest_rew_state_arr)
```

Comment:
Doesn't this mean the reward is recomputed again, in addition to line 265? Also, can't we extract the envs reward instead of computing it in the wrapper?

Reply:
We need to recompute the reward, as we do also in the observations, bc otherwise the last reward is taken. the observations are computed before updating the reward if I am correct.

Reply:
Yeah, that design was a footgun, but not easily fixable now. Let's recompute, maybe adding @cache in the appropriate place will solve performance issues